pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -16,41 +16,187 @@ package packing
+func (p *packingEstimator) getNodeCapacities(filter ...string) []*nodeCapacity {
+	return nodePools
+}
+
+// calculatePackingForPod will calculate the packing for pods starting from
+// start index, it returns the max pods that can be packed and a list of
+// instance types with this pod.
+func (p *packingEstimator) calculatePackingForPod(start int, pods []*v1.Pod) *Packing {
+	podsPacked := []*v1.Pod{}
+	instanceTypesSelected := []*nodeCapacity{}
+	// Get all available instance types for every instance
+	// try to fit as many pods as possible with pods[start] pod.
+
+	// TODO add filters
+	// TODO reserve (Kubelet+ daemon sets) overhead for instance types
+	// TODO count number of pods created on an instance type
+	for _, instanceOption := range p.getNodeCapacities("") {
+		// check how many pods we can fit on this instance type
+		packings, err := p.packingForInstance(instanceOption, start, pods)
+		if err != nil {
+			zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instanceOption.instanceType, err)
+			continue
+		}
+		if len(packings) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if podsMatch(podsPacked, packings) {
+			instanceTypesSelected = append(instanceTypesSelected, instanceOption)
+		} else if len(packings) > len(podsPacked) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			podsPacked = packings
+			instanceTypesSelected = []*nodeCapacity{instanceOption}
+		}
+	}
+	instanceTypeNames := []string{}
+	for _, instance := range instanceTypesSelected {
+		instanceTypeNames = append(instanceTypeNames, instance.instanceType)
+	}
+	return &Packing{Pods: podsPacked, InstanceTypes: instanceTypeNames}
+}
+
+func (p *packingEstimator) packingForInstance(instance *nodeCapacity, start int, pods []*v1.Pod) ([]*v1.Pod, error) {
+	packing := []*v1.Pod{}
+	// start with the largest pod based on resources requested
+	// pods before the start index have been packed
+	for i := start; i < len(pods); i++ {
+		pod := pods[i]
+		if p.isPodPacked(pod) {
+			continue
+		}
+		cpu := cpuFor(pod)
+		memory := memoryFor(pod)
+		if ok := instance.reserveCapacity(*cpu, *memory); !ok {
+			// First we need to find the instance on which we can fit the
+			// largest pods, it can't fit here, return and try with a next
+			// bigger instance type
+			if len(packing) == 0 {
+				return packing, nil
+			}
+			// We have already packed a bigger pod on this instance type, if
+			// we can't pack this pod, check the next pod it might be
+			// smaller in size and help fill the gaps on the instance
+			continue
+		}
+		packing = append(packing, pod)
+	}
+	instance.utilizedCapacity = v1.ResourceList{
+		v1.ResourceCPU:    resource.Quantity{},
+		v1.ResourceMemory: resource.Quantity{},
+	}
+	return packing, nil
+}
+
+func podsMatch(first, second []*v1.Pod) bool {
```

Comment:
Is this just an optimization to avoid DeepEqual()? Potentially premature?

Reply:
Yeah, trying to avoid using reflect and all we need to make sure here is the pods in the two list are the same. DeepEqual seems a little too much for this case.